Bypass the FS check in oe_sgx_get_td during exceptions and oe_abort

mingweishih opened the pull request:
#4387 supports the wrfsbase instruction emulation, which allows the enclave to change FS at runtime. In addition, the PR assumes that the application has to manage the FS by itself; i.e., the application needs to restore the FS to OE's default value when the execution returns to OE runtime. If the application does not restore the FS value, the OE runtime will abort the enclave if the internal API `oe_sgx_get_td` detects the FS value mismatch.

This PR makes a few updates based on the following observations:
- It is not necessary to save and restore the FS value on OCALL enter and return because
  - The `oe_sgx_get_td` is called inthe same function: https://github.com/openenclave/openenclave/blob/0357d6ffcf754fe4dec114ba9a8632dab73f0ca9/enclave/core/sgx/calls.c#L768
  If the FS value is changed at this point, the oe_sgx_get_td should have abort the enclave already.
  - Invoking OCALL should be considered as returning to OE runtime, so the application is expected to restore the FS.

- There are currently two places that need to bypass the FS check in oe_sgx_get_td
   - second-stage exception handler: https://github.com/openenclave/openenclave/blob/0357d6ffcf754fe4dec114ba9a8632dab73f0ca9/enclave/core/sgx/exception.c#L311 given that we cannot expect the application to restore the FS before an exception.
   -  oe_abort which be invoked anywhere, including the second-stage exception handler: https://github.com/openenclave/openenclave/blob/0357d6ffcf754fe4dec114ba9a8632dab73f0ca9/enclave/core/sgx/exception.c#L317
  - The PR introduces `oe_sgx_get_td_no_fs_check` that are used by the above two places to bypass the FS check.
  - The PR also updates `oe_abort` so that it can be called by `oe_sgx_get_td` when FS check fails.
  - In addition, the PR introduces `oe_abort_with_td` to replace some `oe_abort` calls that were previously unavailable because of td was not initialized.






Signed-off-by: Ming-Wei Shih <mishih@microsoft.com>

mingweishih commented:
bors try

anakrish commented on include/openenclave/bits/result.h:
```diff
@@ -377,6 +377,11 @@ typedef enum _oe_result
+    /**
+     * The FS register was in the bad state.
+     */
+    OE_SGX_BAD_FS,
```
Avoid changes to public interface until absolutely needed.

mingweishih replied:
reverted.

anakrish commented on tests/sgx/wrfsbase/enc/enc.c:
```diff
@@ -8,34 +8,132 @@
     OE_TEST(new_fs == temp_td);
-    OE_TEST(recovered_fs == old_fs);
+    OE_TEST(fs_after_cpuid == temp_td);
+    OE_TEST(fs_after_exception == original_fs);
```
It is counter-intuitive that when control resumes after an exception, the FS value is not the value just before exception.

mingweishih replied:
updated.

anakrish reviewed (changes requested):
This PR needs further clarification

1.  The published interface (oe_result_t) is changed. There needs to be a strong rationale for that.
2.  The value of FS after an exception may not be the value just before an exception. Exceptions can be triggered at any point in code (e.g via SIG35). It would lead to undefined behavior if, upon continuation from the exception, the register environment is changed.
3. This PR removes code that was just merged in 6 days ago: https://github.com/openenclave/openenclave/commit/f70c404dec73b760ae84e64f5343c95cd1c0597e. It is not clear why that code is no longer valid. If these changes are experimental, then I'd prefer that we submit this PR to a separate dev branch and then submit a PR to OE's main branch only after the changes have stabilized.

mingweishih commented:
bors try

mingweishih commented:
> This PR needs further clarification
> 
> 1. The published interface (oe_result_t) is changed. There needs to be a strong rationale for that.

Reverted

> 2. The value of FS after an exception may not be the value just before an exception. Exceptions can be triggered at any point in code (e.g via SIG35). It would lead to undefined behavior if, upon continuation from the exception, the register environment is changed.

Updated the implementation.

> 3. This PR removes code that was just merged in 6 days ago: [f70c404](https://github.com/openenclave/openenclave/commit/f70c404dec73b760ae84e64f5343c95cd1c0597e). It is not clear why that code is no longer valid. If these changes are experimental, then I'd prefer that we submit this PR to a separate dev branch and then submit a PR to OE's main branch only after the changes have stabilized.

Added explanation in the PR description. Now the PR is in the mystikos_upstream branch. I'll hold this PR until Mystikos pipeline against the branch passes.

mingweishih commented:
bors try

CyanDevs commented:
bors try

CyanDevs commented:
bors try

mingweishih commented:
bors try

mingweishih commented:
bors try

mingweishih commented:
bors try

mingweishih commented:
bors try-

mingweishih commented:
bors try

anakrish commented on enclave/core/sgx/calls.c:
```diff
@@ -1246,9 +1233,54 @@ void oe_abort(void)
+    oe_sgx_td_t* td = oe_sgx_get_td_no_fs_check();
+
+    /* It is unlikely that td is invalid. If this is the case, we cannot
+     * call _abort to exit the enclave. Instead, we intentionally trigger
```
Nice idea to immediately exit the enclave via page fault. 
Is there a way to also get other threads to immediately exit the enclave when one thread aborts?

mingweishih replied:
This may require some signal-like mechanisms. We can think about it later.

anakrish commented on enclave/core/sgx/exception.c:
```diff
@@ -308,7 +308,10 @@ int _emulate_illegal_instruction(sgx_ssa_gpr_t* ssa_gpr)
+    /* Bypass the FS check given that we cannot except the application
+     * to restore the FS before an exception (if the application has
+     * modified the FS) */
+    oe_sgx_td_t* td = oe_sgx_get_td_no_fs_check();
```
Does linux follow the same behavior for exception handler in terms for FS/GS?

mingweishih replied:
Linux kernel does not use FS (thread local storage is not supported in kernel space). For GS, the Linux kernel will save and restore the GS value when entering and exiting the kernel.

anakrish commented on tests/sgx/wrfsbase/enc/enc.c:
```diff
@@ -8,34 +8,170 @@
 
+    if (!simulation_mode)
+    {
+        OE_TEST(original_fs_after_exception == original_fs);
```
These invariants are much simpler to reason about and code against.

anakrish reviewed (approved):
LGTM.
As a future task, consider if there is a way to abort all the threads immediately upon aborting one thread. This is something that we haven't figured out for quite sometime.

mingweishih commented:
bors try

mingweishih commented:
bors try-

mingweishih commented:
bors try

mingweishih commented:
bors try

radhikaj reviewed (approved):
bors r+
